src/boxplot.rs

```diff
@@ -0,0 +1,241 @@
+use super::{AsMatrix, matrix_to_array, generate_list, GraphMaker};
+use std::fmt::Write;
+
+/// Draw a box and whisker plot
+///
+/// [See Matplotlib's documentation](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.boxplot.html)
+///
+/// # Examples
+///
+/// ## Basic box plot
+///
+/// ```
+/// use plotpy::{Boxplot, AsMatrix, Plot, StrError};
+///
+/// fn main() -> Result<(), StrError> {
+///     // data
+///     let x = vec![vec![1,2,3,4,5],
+///                  vec![2,3,4,5,6],
+///                  vec![3,4,5,6,7],
+///                  vec![4,5,6,7,8],
+///                  vec![5,6,7,8,9],
+///                  vec![6,7,8,9,10]];
+/// 
+///     // x ticks and labels
+///     let ticks: Vec<_> = (1..=x.size().1).into_iter().collect();
+///     let labels = ["x1", "x2", "x3", "x4", "x5"];
+/// 
+///     // boxplot object and options
+///     let mut boxes = Boxplot::new();
+///     boxes.set_vertical(true)
+///          .draw(&x);
+///
+///     // save figure
+///     let mut plot = Plot::new();
+///     plot.add(&boxes).set_title("boxplot")
+///         .set_ticks_x_labels(&ticks, &labels)
+///         .set_labels("x", "y")  // x-axis and y-axis label
+///         .save("/tmp/plotpy/doc_tests/doc_boxplot_1.svg")?;
+///     Ok(())
+/// }
+/// ```
+/// 
+/// ![doc_boxplot_1.svg](https://raw.githubusercontent.com/cpmech/plotpy/main/figures/doc_boxplot_1.svg)
+/// 
+/// 
+/// ## More examples
+/// 
+/// See also integration test in the **tests** directory.
+pub struct Boxplot {
+    label: String,              // Name of this box in the legend
+    symbol: Option<String>,     // The default symbol for flier (outlier) points.
```

Comment:
Hi, I thought a lot when deciding whether to use `Option`. `symbol: Option<String>` is necessary here, we cannot have an empty string as the default simply, because `None` means the default value `b+` in matplotlib boxplot, while empty strings means no flier points on the boxplot, so we'd better use `Option<String>` here.

Reply:
Thanks. I still don't see the need for Option here. I've actually re-implemented it already with:

```rust
if self.symbol != "" {
    write!(&mut opt, ",sym=r'{}'", self.symbol).unwrap();
}
```

Reply:
In this implementation, the boxplot will always show the flier, even if the symbol is an empty string. we have no way to disable the flier, what should we do if we don't want the flier ?

Reply:
Got it.

Reply:
Could you set the defalut value of `symbol` to `b+` ? This is the default value used by matplotlib:
```text
The default symbol for flier points. An empty string ('') hides the fliers. If [None], then the fliers default to 'b+'. More control is provided by the flierprops parameter.
```
If we set it to empty string, we must set it to 'b+' every time we plot a boxplot. By convention, the default value should be the mostly used value. And this is also why I used `Option<String>` before, because I can set to `None`, then the default `b+` will be used by matplotlib when plotting.

Reply:
We are already using the default symbol by Matplotlib. In fact, we are using all the defaults. The idea is "*do not specify anything unless the user wants to select something different than the default*". For instance, the `options` function does nothing unless some value is given:

```rust
let mut opt = String::new();
if self.symbol != "" {
    write!(&mut opt, ",sym=r'{}'", self.symbol).unwrap();
}
if self.horizontal {
    write!(&mut opt, ",vert=False").unwrap();
}
if self.whisker != None {
    write!(&mut opt, ",whis={}", self.whisker.unwrap()).unwrap();
}
if self.positions.len() > 0 {
    write!(&mut opt, ",positions=positions").unwrap();
}
if self.width != None {
    write!(&mut opt, ",widths={}", self.width.unwrap()).unwrap();
}
if self.extra != "" {
    write!(&mut opt, ",{}", self.extra).unwrap();
}
```

In this case, the actual Matplotlib command will be simply:

```python
p=plt.boxplot(x)
```

And the default Matplolib values are used.

If the user always wants to run Boxplot with his own specifications, the user could create a wrapping function (or a derived structure).

Reply:
Oh, now I understand, Matplotlib has an option to disable the symbols altogether, right?

Reply:
I've implemented the function `set_no_fliers` in Boxplot in case the user doesn't want fliers (available in version **1.9.1**). The default symbol is "untouched," and Matplolib will draw (in my Matplotlib's version 3.6) an open black circle by default (due to `p=plt.boxplot(x)`).

This case highlights an issue I see with Matplotlib, things are sometimes confusing! For instance, using the symbol marker/color to enable disable markers is confusing. Therefore, `plotpy` seeks to be more explicit about configuration parameters. That's why I prefer to have a function `set_no_fliers`, which is "totally decoupled" from the symbol configuration.

Reply:
I think I know what you mean. However, now we cannot disable the flier, no matter the `symbol` is empty or not.  when it's empty, then the command is `p=plt.boxplot(x)`, the filers will be displayed as circles. when it's non-empty, the fliers will be the shape we set.  We cannot disable the fliers in either case. I meant to disable them using `Option(String::new())`, because an emtpy string in matploblib boxplot will disable the flier, and `None` will enable the flier as circles. Now we use `String`, I think we should add a bool parameter called `showfliers` to enable or disable the filer. Hope I don't confuse you.
![Uploading image.png…]()

Reply:
I did some testing with version 1.9.1, and we can disable the fliers. Please check it out and correct me if I'm wrong.

Yes, I've seen this `showfliers` function, but it doesn't work! (please check it out using Python directly)

Reply:
`showfilers` works in the latest matplotlib, perhaps the issue is fixed now. In case we have `set_no_fliers` now, we can also control the fliers.

Reply:
Here is an example:

```rust
let data = vec![
    vec![1, 2, 3, 4, 5, 10],          // A
    vec![2, 3, 4, 5, 6, 7, 8, 9, 10], // B
    vec![3, 4, 5, 6, 10],             // C
];

let ticks = [1, 2, 3];
let labels = ["A", "B", "C"];

let mut b1 = Boxplot::new();
b1.draw(&data); // the default in my system is a black circle

let mut b2 = Boxplot::new();
b2.set_symbol("b+").draw(&data);

let mut b3 = Boxplot::new();
b3.set_symbol("b+").set_no_fliers(true).draw(&data);

let mut plot = Plot::new();
plot.set_vertical_gap(0.0)
    .set_subplot(3, 1, 1)
    .add(&b1)
    .set_subplot(3, 1, 2)
    .add(&b2)
    .set_subplot(3, 1, 3)
    .add(&b3)
    .set_ticks_x_labels(&ticks, &labels)
    .set_figure_size_points(300.0, 450.0)
    .save("/tmp/plotpy/test_boxplot.svg")
```

![test_boxplot](https://github.com/user-attachments/assets/0eaebcaf-4595-4766-8d65-351c1f8badaa)